File: app/src/main/java/jp/juggler/subwaytooter/dialog/AccountPicker.java
package jp.juggler.subwaytooter.dialog;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.concurrent.atomic.AtomicBoolean;

import jp.juggler.subwaytooter.ActMain;
import jp.juggler.subwaytooter.R;
import jp.juggler.subwaytooter.table.AcctColor;
import jp.juggler.subwaytooter.table.SavedAccount;
import jp.juggler.subwaytooter.util.Utils;

import android.widget.LinearLayout;
import android.widget.TextView;

public class AccountPicker {
	
	public interface AccountPickerCallback {
		void onAccountPicked( @NonNull SavedAccount ai );
	}
	
	
	public static void pick(
		@NonNull AppCompatActivity activity
		, boolean bAllowPseudo
		, boolean bAuto
		, String message
		, @NonNull final AccountPickerCallback callback
	){
		
		ArrayList< SavedAccount > account_list = SavedAccount.loadAccountList( activity,ActMain.log );
		SavedAccount.sort( account_list );
		pick( activity, bAllowPseudo, bAuto, message, account_list, true,callback ,null );
	}
	
	public static void pick(
		@NonNull AppCompatActivity activity
		, boolean bAllowPseudo
		, boolean bAuto
		, String message
		, @NonNull final AccountPickerCallback callback
		, @Nullable final DialogInterface.OnDismissListener dismiss_callback
	){
		ArrayList< SavedAccount > account_list = SavedAccount.loadAccountList( activity,ActMain.log );
		pick( activity, bAllowPseudo, bAuto, message, account_list, true, callback ,dismiss_callback );
	}
	
	public static void pick(
		@NonNull AppCompatActivity activity
		, boolean bAllowPseudo
		, boolean bAuto
		, String message
		, @NonNull final ArrayList< SavedAccount > account_list
		, @NonNull final AccountPickerCallback callback
	){
		pick( activity, bAllowPseudo, bAuto, message, account_list, false, callback ,null );
	}
	
	private static void pick(
		@NonNull AppCompatActivity activity
		, boolean bAllowPseudo
		, boolean bAuto
		, String message
		, @NonNull final ArrayList< SavedAccount > account_list
	    ,boolean bSort
		, @NonNull final AccountPickerCallback callback
		, @Nullable final DialogInterface.OnDismissListener dismiss_callback
	){
		if( account_list.isEmpty() ){
			Utils.showToast( activity, false, R.string.account_empty );
			return;
		}
		
		if( ! bAllowPseudo ){
			ArrayList< SavedAccount > tmp_list = new ArrayList<>();
			for( SavedAccount a : account_list ){
				if( a.isPseudo() ) continue;
				tmp_list.add( a );
			}
			account_list.clear();
			account_list.addAll( tmp_list );
			if( account_list.isEmpty() ){
				Utils.showToast( activity, false, R.string.not_available_for_pseudo_account );
				return;
			}
		}
		
		if(bSort){
			SavedAccount.sort( account_list );
		}
		
		if( bAuto && account_list.size() == 1 ){
			callback.onAccountPicked( account_list.get( 0 ) );
			return;
		}
		
		
		
		@SuppressLint("InflateParams") View viewRoot = activity.getLayoutInflater().inflate( R.layout.dlg_account_picker, null, false );

		final Dialog dialog = new Dialog( activity );
		final AtomicBoolean isDialogClosed = new AtomicBoolean( false );
		
		if( dismiss_callback != null ){
			dialog.setOnDismissListener( dismiss_callback );
		}

		dialog.setContentView( viewRoot );
		if( ! TextUtils.isEmpty( message ) ){
			TextView tv = (TextView) viewRoot.findViewById( R.id.tvMessage );
			tv.setVisibility( View.VISIBLE );
			tv.setText( message );
		}
		viewRoot.findViewById( R.id.btnCancel ).setOnClickListener( new View.OnClickListener() {
			@Override public void onClick( View v ){
				isDialogClosed.set(true);
				dialog.cancel();
			}
		} );
		dialog.setCancelable( true );
		dialog.setCanceledOnTouchOutside( true );
		dialog.setOnCancelListener( new DialogInterface.OnCancelListener() {
			@Override public void onCancel( DialogInterface dialog ){
				isDialogClosed.set(true);
			}
		} );
		
		
		float density = activity.getResources().getDisplayMetrics().density;
		
		
		LinearLayout llAccounts = (LinearLayout) viewRoot.findViewById( R.id.llAccounts );
		int pad_se = (int) ( 0.5f + 12f * density );
		int pad_tb = (int) ( 0.5f + 6f * density );
		
		for( SavedAccount a : account_list ){
			LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams( LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT );
			
			AcctColor ac = AcctColor.load( a.acct );
			
			Button b = new Button( activity );
			
			if( AcctColor.hasColorBackground( ac ) ){
				b.setBackgroundColor( ac.color_bg );
			}else{
				b.setBackgroundResource( R.drawable.btn_bg_transparent );
			}
			if( AcctColor.hasColorForeground( ac ) ){
				b.setTextColor( ac.color_fg );
			}
			
			b.setPaddingRelative( pad_se, pad_tb, pad_se, pad_tb );
			b.setGravity( Gravity.START | Gravity.CENTER_VERTICAL );
			b.setAllCaps( false );
			b.setLayoutParams( lp );
			b.setMinHeight( (int) ( 0.5f + 32f * density ) );
			b.setText( AcctColor.hasNickname( ac ) ? ac.nickname : a.acct );
			
			final SavedAccount _a = a;
			b.setOnClickListener( new View.OnClickListener() {
				@Override public void onClick( View v ){
					isDialogClosed.set(true);
					callback.onAccountPicked( _a );
					dialog.dismiss();
				}
			} );
			llAccounts.addView( b );
		}
		
		dialog.show();
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/dialog/AccountPicker.kt
package jp.juggler.subwaytooter.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.DialogInterface
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.RelativeSizeSpan
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import jp.juggler.subwaytooter.R
import jp.juggler.subwaytooter.table.*
import jp.juggler.util.data.notEmpty
import jp.juggler.util.log.LogCategory
import jp.juggler.util.log.showToast
import jp.juggler.util.ui.dismissSafe
import jp.juggler.util.ui.getAdaptiveRippleDrawableRound
import org.jetbrains.anko.textColor
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

private val log = LogCategory("pickAccount")

@SuppressLint("InflateParams")
suspend fun AppCompatActivity.pickAccount(
    bAllowPseudo: Boolean = false,
    bAllowMisskey: Boolean = true,
    bAllowMastodon: Boolean = true,
    bAuto: Boolean = false,
    message: String? = null,
    accountListArg: List<SavedAccount>? = null,
    dismissCallback: (dialog: DialogInterface) -> Unit = {},
    extraCallback: (LinearLayout, Int, Int) -> Unit = { _, _, _ -> },
): SavedAccount? {
    val activity = this
    var removeMastodon = 0
    var removedMisskey = 0
    var removedPseudo = 0

    fun SavedAccount.checkMastodon() = when {
        !bAllowMastodon && !isMisskey -> ++removeMastodon
        else -> 0
    }

    fun SavedAccount.checkMisskey() = when {
        !bAllowMisskey && isMisskey -> ++removedMisskey
        else -> 0
    }

    fun SavedAccount.checkPseudo() = when {
        !bAllowPseudo && isPseudo -> ++removedPseudo
        else -> 0
    }

    val accountList = accountListArg
        ?: daoSavedAccount.loadAccountList()
            .filter { 0 == it.checkMastodon() + it.checkMisskey() + it.checkPseudo() }
            .sortedByNickname()

    if (accountList.isEmpty()) {

        val sb = StringBuilder()

        if (removedPseudo > 0) {
            sb.append(activity.getString(R.string.not_available_for_pseudo_account))
        }

        if (removedMisskey > 0) {
            if (sb.isNotEmpty()) sb.append('\n')
            sb.append(activity.getString(R.string.not_available_for_misskey_account))
        }
        if (removeMastodon > 0) {
            if (sb.isNotEmpty()) sb.append('\n')
            sb.append(activity.getString(R.string.not_available_for_mastodon_account))
        }

        if (sb.isEmpty()) {
            sb.append(activity.getString(R.string.account_empty))
        }

        activity.showToast(false, sb.toString())
        return null
    }

    if (bAuto && accountList.size == 1) {
        return accountList[0]
    }

    return suspendCoroutine { continuation ->
        val viewRoot = layoutInflater.inflate(R.layout.dlg_account_picker, null, false)
        val dialog = Dialog(activity)
        val isResumed = AtomicBoolean(false)

        dialog.setOnDismissListener {
            dismissCallback(it)
            if (isResumed.compareAndSet(false, true)) {
                continuation.resume(null)
            }
        }

        dialog.setContentView(viewRoot)
        if (message != null && message.isNotEmpty()) {
            val tv = viewRoot.findViewById<TextView>(R.id.tvMessage)
            tv.visibility = View.VISIBLE
            tv.text = message
        }

        viewRoot.findViewById<View>(R.id.btnCancel).setOnClickListener {
            dialog.cancel()
        }

        dialog.setCancelable(true)
        dialog.setCanceledOnTouchOutside(true)

        val density = activity.resources.displayMetrics.density

        val llAccounts: LinearLayout = viewRoot.findViewById(R.id.llAccounts)
        val padX = (0.5f + 12f * density).toInt()
        val padY = (0.5f + 6f * density).toInt()

        for (a in accountList) {
            val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )

            val ac = daoAcctColor.load(a)

            val b = AppCompatButton(activity)

            if (daoAcctColor.hasColorBackground(ac)) {
                b.background = getAdaptiveRippleDrawableRound(
                    activity,
                    ac.colorBg,
                    ac.colorFg
                )
            } else {
                b.setBackgroundResource(R.drawable.btn_bg_transparent_round6dp)
            }
            if (daoAcctColor.hasColorForeground(ac)) {
                b.textColor = ac.colorFg
            }

            b.setPaddingRelative(padX, padY, padX, padY)
            b.gravity = Gravity.START or Gravity.CENTER_VERTICAL
            b.isAllCaps = false
            b.layoutParams = lp
            b.minHeight = (0.5f + 32f * density).toInt()

            val sb = SpannableStringBuilder(ac.nickname)
            try {
                val status = daoAccountNotificationStatus.load(a.acct)
                val lastNotificationError = status?.lastNotificationError?.notEmpty()
                val lastSubscriptionError = status?.lastSubscriptionError?.notEmpty()
                (lastNotificationError ?: lastSubscriptionError)?.let { message ->
                    sb.append("\n")
                    val start = sb.length
                    sb.append(message)
                    val end = sb.length
                    sb.setSpan(
                        RelativeSizeSpan(0.7f),
                        start,
                        end,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
            } catch (ex: Throwable) {
                log.e(ex, "can't get notification status for ${a.acct}")
            }
            b.text = sb

            b.setOnClickListener {
                if (isResumed.compareAndSet(false, true)) {
                    continuation.resume(a)
                }
                dialog.dismissSafe()
            }
            llAccounts.addView(b)
        }

        extraCallback(llAccounts, padX, padY)

        dialog.show()
    }
}
